twoliter/src/project/migrate/mod.rs

```diff
@@ -0,0 +1,553 @@
+    pub fn migrate_project(&self, content: &str, from: u32, to: u32) -> Result<String> {
+        if from == to {
+            debug!("No migration needed: versions are the same ({})", from);
+            return Ok(content.to_string());
+        }
+
+        info!("Migrating project from version {} to version {}", from, to);
+
+        // Try direct migration first
+        if let Some(migrator) = self.find_migrator(from, to) {
+            let result = migrator
+                .migrate(content)
+                .with_context(|| format!("Failed to migrate from version {} to {}", from, to))?;
+            info!(
+                "Successfully migrated project from version {} to version {}",
+                from, to
+            );
+            return Ok(result);
+        }
+
+        // If no direct migration, try step-by-step migration
+        self.migrate_step_by_step(content, from, to)
+    }
+
+    /// Perform step-by-step migration when no direct path exists.
+    fn migrate_step_by_step(&self, content: &str, from: u32, to: u32) -> Result<String> {
+        let mut current_content = content.to_string();
+        let mut current_version = from;
+
+        if from < to {
+            // Forward migration (step by step)
+            while current_version < to {
+                let next_version = current_version + 1;
+                if let Some(migrator) = self.find_migrator(current_version, next_version) {
+                    current_content = migrator.migrate(&current_content).with_context(|| {
+                        format!(
+                            "Failed to migrate from version {} to {}",
+                            current_version, next_version
+                        )
+                    })?;
+                    current_version = next_version;
+                } else {
+                    anyhow::bail!(
+                        "No migration path found from version {} to version {}",
+                        current_version,
+                        next_version
+                    );
+                }
+            }
+        }
+
+        Ok(current_content)
+    }
+}
+
+/// Detect the schema version of a project file without fully parsing it.
+pub fn detect_schema_version(content: &str) -> Result<u32> {
+    // Parse just enough to get the schema version
+    let value: toml::Value = toml::from_str(content).context("Failed to parse TOML content")?;
+
+    let version = value
+        .get("schema-version")
+        .or_else(|| value.get("schema_version"))
+        .context("No schema-version field found in project file")?
+        .as_integer()
+        .context("schema-version field is not an integer")?;
+
+    Ok(version as u32)
+}
+
+/// Migrate project content to the target version.
+/// This is the main function used by the project loading system to automatically
+/// handle migration when loading Twoliter.toml files.
+pub fn migrate_project_content(content: &str, target_version: u32) -> Result<String> {
```

Comment:
Per my comment above, regardless of how we change the migrator's intermediate representation, I think we should return something much more structured than `String` here.